src/plugin.rs

```diff
@@ -0,0 +1,216 @@
+use std::str::FromStr;
+use std::sync::Arc;
+
+use pyo3::{prelude::*, types::PyType};
+
+use crate::{DecryptError, EncryptError, IdentityError, RecipientError};
+
+/// Hack, because the orphan rule would prevent us from deriving a
+/// foreign trait on a foreign object. Instead, define a newtype.
+///
+/// Inner type is PyAny, because we do duck-typing at runtime, and
+/// declaring a protocol in the type stubs.
+#[derive(Clone)]
+pub(crate) struct PyCallbacks(Py<PyAny>);
+
+impl PyCallbacks {
+    fn new(inner: Bound<'_, PyAny>) -> PyResult<Self> {
+        {
+            for method_name in [
+                "display_message",
+                "confirm",
+                "request_public_string",
+                "request_passphrase",
+            ] {
+                // I would use `hasattr`, but it doesn't throw, and
+                // constructing pretty AttributeErrors is incredibly
+                // hard.
+                //
+                // XXX: Should we even throw it? Better ask for
+                // forgiveness than permission, and we're panicking
+                // when things get called. And static typecheckers
+                // will catch a protocol violation.
+                inner.getattr(method_name)?;
+            }
```

Comment:
Yeah, thinking about it a bit more, I'd be okay with removing this -- I agree with your comment that we don't accomplish very much by enforcing attribute presence here.